generate.go

```diff
@@ -0,0 +1,179 @@
+// Copyright 2021 Mineiros GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package terramate
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/hashicorp/hcl/v2/hclsyntax"
+	"github.com/hashicorp/hcl/v2/hclwrite"
+	"github.com/madlambda/spells/errutil"
+	"github.com/mineiros-io/terramate/hcl"
+)
+
+const (
+	// GeneratedTfFilename is the name of the terramate generated tf file.
+	GeneratedTfFilename = "_gen_terramate.tm.tf"
+
+	// GeneratedCodeHeader is the header added on all generated files.
+	GeneratedCodeHeader = "// GENERATED BY TERRAMATE: DO NOT EDIT\n\n"
+)
+
+// Generate will walk all the directories starting from basedir generating
+// code for any stack it finds as it goes along
+//
+// It will return an error if it finds any invalid terramate configuration files
+// of if it can't generate the files properly for some reason.
+//
+// The provided basedir must be an absolute path to a directory.
+func Generate(basedir string) error {
```

Comment:
As generate could partially work (generate until error), it'd be cool if you return the list of generated files (alongside the error, it that's the case) so we can have a nice output in the cli, like:

```
$ terramate generate
2021/12/11 21:22:23 failed to generate code for stack "/home/i4k/src/mineiros/tf-stacks/envs/prod/wg": parsing config "terramate.tm.hcl": malformed terramate block: invalid attribute "after": parsing config: malformed terramate block: invalid attribute "after"

Generated files:
[+] envs/prod/vpn/_gen_terramate.tm.tf
[+] envs/prod/vpc/_gen_terramate.tm.tf
[+] envs/dev/vpn/_gen_terramate.tm.tf

Exited with error 1
$
```

Reply:
Thanks makes perfect sense, will create a task to address this ASAP :100: